types/underscore/index.d.ts

```diff
@@ -106,22 +107,28 @@ declare module _ {
         (prev: TResult, curr: T, key: string, list: Dictionary<T>): TResult;
     }
 
+    type TypeOfList<V> = V extends _.List<infer T> ? T : never;
+
+    type TypeOfDictionary<V> = V extends _.Dictionary<infer T> ? T : never;
+
+    type TypeOfCollection<V> = V extends _.Collection<infer T> ? T : never;
+
+    type _ChainSingle<V> = _Chain<TypeOfCollection<V>, V>;
```

Comment:
Is the use of `never` safe? It's just a guess, but my first impression is that the following will break:

```ts
_.chain([1, 2, 3]).find(n => n % 2).value();
```

Reply:
Happily `find` specifically has an extensive set of tests in a form that I actually based most of the tests I wrote off of (see `namespace TestFind` in `underscore-tests.ts`, so I can confidently say that the example you provide works 😄 

What happens there is that you end up with a `_Chain<T, V>` (where `T` represents the type of items in a collection and `V` represents the wrapped value) that is `_Chain<never, number | undefined>`, and since `value` is defined as `value(): V;` the result ends up being `number | undefined`.

What the `never` does get in the way of is using subsequent functions that depend on the collection item type `T`; for example, the second find call below would get an `n` of `never` since the chain is not wrapping a collection type where having a collection item type makes sense.

```TS
_.chain([1, 2, 3]) // result is _Chain<number, number[]>
    .find(n => n % 2) // result is _Chain<never, number | undefined>
    .find(n => n % 2) // This seems like it should be an error, but apparently doing math on a `never` is allowed. It would prevent most other operations though.
    .value();
```

(Sigh) Or at least it would if `_Chain.find` didn't have a few issues with its present declaration including its own local redeclaration of the `T` generic and a union of iterator types that makes TS less good at automatically inferring the types that should be applied to callbacks.